qt/aqt/table.py

```diff
@@ -0,0 +1,1135 @@
+        self._last_refresh = time.time()
+        self.dataChanged.emit(top_left, bottom_right)  # type: ignore
+
+    # Item Interface
+    ######################################################################
+
+    # Get metadata
+
+    def is_empty(self) -> bool:
+        return not self._items
+
+    def len_rows(self) -> int:
+        return len(self._items)
+
+    def len_columns(self) -> int:
+        return len(self._state.active_columns)
+
+    # Get items (card or note ids depending on state)
+
+    def get_item(self, index: QModelIndex) -> Item:
+        return self._items[index.row()]
+
+    def get_items(self, indices: List[QModelIndex]) -> List[Item]:
+        return [self.get_item(index) for index in indices]
+
+    def get_card_ids(self, indices: List[QModelIndex]) -> List[CardId]:
+        return self._state.get_card_ids(self.get_items(indices))
+
+    def get_note_ids(self, indices: List[QModelIndex]) -> List[NoteId]:
+        return self._state.get_note_ids(self.get_items(indices))
+
+    # Get row numbers from items
+
+    def get_item_row(self, item: Item) -> Optional[int]:
+        for row, i in enumerate(self._items):
+            if i == item:
+                return row
+        return None
+
+    def get_item_rows(self, items: Sequence[Item]) -> List[int]:
+        rows = []
+        for row, i in enumerate(self._items):
+            if i in items:
+                rows.append(row)
+        return rows
+
+    def get_card_row(self, card_id: CardId) -> Optional[int]:
+        return self.get_item_row(self._state.get_item_from_card_id(card_id))
+
+    # Get objects (cards or notes)
+
+    def get_card(self, index: QModelIndex) -> Optional[Card]:
+        """Try to return the indicated, possibly deleted card."""
+        try:
+            return self._state.get_card(self.get_item(index))
+        except NotFoundError:
+            return None
+
+    def get_note(self, index: QModelIndex) -> Optional[Note]:
+        """Try to return the indicated, possibly deleted note."""
+        try:
+            return self._state.get_note(self.get_item(index))
+        except NotFoundError:
+            return None
+
+    # Table Interface
+    ######################################################################
+
+    def toggle_state(self, context: SearchContext) -> ItemState:
+        self.beginResetModel()
+        self._state = self._state.toggle_state()
+        self.search(context)
+        return self._state
+
+    # Rows
+
+    def search(self, context: SearchContext) -> None:
+        self.begin_reset()
+        try:
+            gui_hooks.browser_will_search(context)
+            if context.ids is None:
+                context.ids = self._state.find_items(context.search, context.order)
+            gui_hooks.browser_did_search(context)
+            self._items = context.ids
+        finally:
+            self.end_reset()
+
+    def reverse(self) -> None:
```

Comment:
We've been emptying the cache here before, but I don't see why that should be necessary.

Reply:
I think this was to ensure that when a card was changed, we didn't serve up stale results based on the old cached version we had. But you're right, we probably don't need it now